Feature: Local cache for op items

odelucca opened the pull request:
## ☕ Purpose

After I started using #13, I've noticed that it takes too long if every time I fetch passwords my computer needs to go to the 1Password server to fetch them.

To fix this, I've created this PR, which saves a local cache containing the result of `op list-items`. The cache is available for 6 hours :)

This PR is based on #13, so some of the code here may looks different but it is actually the implementation of #13. As soon as that PR is merged, it would be more clear to see the diffs.

## 🧐 Checklist

- [x] Adds a function to save the cache
- [x] Adds a condition to use cache instead of fetch, if available
- [x] Adds a condition to remove the cache if it is older than  6hours

camspiers commented:
While I love the idea of caching, putting all of my incredibly sensitive passwords in an unencrypted file in a directory that every program running on my computer has read access to seems like a really bad idea, and not something I would be comfortable with from a security perspective.

odelucca commented:
@camspiers the file contains only the title and ID of the account :) it does not contain the password itself.

The password is fetched by the plug-in afterwards, when you select which account you want to get

odelucca commented:
Just to give more details, the data I'm caching is the return of `op list-items` after being filtered by `jq`, which results only in tuples containing: `(Account title, Account ID)`

This is not sensitive data, so there is no need to encrypt it.

Even if someone stole your file, they would need your 1Password password to fetch the actual data for your account. So, in my opinion there is no need for any extra security step here

camspiers commented:
Oh right. Excellent!

odelucca commented:
@yardnsm can you take a look at this please? :D thanks!

camspiers commented:
btw, apologies for the false assumption above. This feature is going to be really valuable. Thanks!

yardnsm commented on scripts/main.sh:
```diff
@@ -44,7 +46,16 @@ op_get_session() {
 }
 
 get_op_items() {
+  local cache_file="/tmp/tmux-op-items"
```
Should be at the top as a constant, like the `TMP_TOKEN_FILE` variable.

yardnsm commented on scripts/main.sh:
```diff
@@ -44,7 +46,16 @@ op_get_session() {
 get_op_items() {
+  local cache_file="/tmp/tmux-op-items"
 
+  if [ -e $cache_file ]; then
```
Use the "newer" variant:

```suggestion
  if [[ -e $cache_file ]]; then
```

yardnsm commented on README.md:
```diff
@@ -77,8 +77,9 @@ sessions expires automatically after 30 minutes of inactivity.
+[sudolikeaboss](https://github.com/ravenac95/sudolikeaboss), by default it is required to set the value of the
+`website` field for each login item with the value of `sudolikeaboss://local`. To override this behavior and provide
+customized filtering, see configuration options `@1password-enabled-url-filter` and `@1password-items-jq-filter` below.
 
```
I would add a "Security" section to the README explaining that we're saving some data (and what data) from the vaults to the disk.

yardnsm commented on scripts/main.sh:
```diff
@@ -58,18 +69,48 @@ get_op_items() {
+  local items=$1
+  local cache_file="/tmp/tmux-op-items"
+
+  if ! [ -e $cache_file ]; then
```
Same here:

```suggestion
  if ! [[ -e $cache_file ]]; then
```

yardnsm commented on scripts/main.sh:
```diff
@@ -58,18 +69,48 @@ get_op_items() {
+    local seconds_since_last_update="$(($now-$last_update))"
+
+    # Remove cache file if last cache was from 6h ago
+    if [[ $seconds_since_last_update < 21600 ]]; then
```
Also move `21600` into a constant as the top

odelucca commented:
@yardnsm nice idea! I'm going to search for a proper keybinding to use!

About `1pass`, in my opinion it is an awesome tool, but too hard to setup (I tried to setup it many times and there is a bunch of issues with it) and it doesn't adds too much value to `op` (in my opinion). 

So, this could be a good alternative to those who won't use `1pass`

odelucca commented:
@yardnsm I'm going to take a look on your suggestions on the following days.

Also, I've noticed another problem. Since we're just asking for user password during the `fetch_items`, if the user session has expired after selecting a cached item the user would not be able to fetch the password itself.

I must also fix this by asking for the password during afterwards if the session has expired

odelucca commented:
@yardnsm I've updated the code :) can you please review it and merge with master?

I've also updated the `tmux-token` path (from `$HOME` to `/tmp`), to match the pattern of storing temporary data on the right place (`$HOME` should not be used for tmp data)

The only thing that I wasn't able to do was add a condition to login the user if the session expired. `op` doesn't provide a command to validate the current session, so I reduced the `CACHE_TTL` to 30 minutos in order to match the session token TTL.

yardnsm commented on plugin.tmux:
```diff
@@ -28,9 +28,12 @@ main() {
   tmux bind-key "$opt_key" \
     run "tmux split-window -l 10 \"$CURRENT_DIR/scripts/main.sh '#{pane_id}'\""
+
+  tmux bind-key "$clear_key" run "$CURRENT_DIR/scripts/clear.sh"
```
Urghhh the `clear.sh` has a duplicated logic as in `main.sh`. I suggest making this line as:

```suggestion
  tmux bind-key "$clear_key" run "$CURRENT_DIR/scripts/main.sh clear-cache"
```

And then, in `main.sh`, check the value of the given argument. If it is `"clear-cache"`, then clear the cache. Else, run the usual flow.

yardnsm commented on scripts/main.sh:
```diff
@@ -84,6 +93,23 @@ get_op_items() {
+    local now="$(date +%s)"
+    local seconds_since_last_update="$(($now-$last_update))"
+
+    # Remove cache file if last cache was from 6h ago
```
It's not 6h anymore

odelucca commented:
@yardnsm thanks for the review, I'm going to take a look on your comments.

odelucca commented:
@yardnsm I've fixed your comments, and also fixed a bug.

The cache was being automatically cleared before `fetch_items`. Now it automatically clears the cache, before fetching items, if the cache is old enought. :)

Waiting for you review!

yardnsm commented on scripts/main.sh:
```diff
@@ -127,6 +159,22 @@ get_op_item_password() {
 # ------------------------------------------------------------------------------
 
 main() {
```
Please move the `main` function to the bottom.

odelucca commented:
@yardnsm done :) Waiting for you approval

yardnsm commented:
I've merged #18, which continues #13. You can revert the changes from #13 and it should be fine 👌

odelucca commented:
Great! Thanks.

I'm going to revert it in the next few days and get in touch with you ;)

odelucca commented:
@yardnsm sorry for the late reply. I've just finished it and I'm merging right now ;) Thanks!

odelucca commented:
@yardnsm will this branch ever be merged? :)

tapayne88 commented:
This looks good, great work @odelucca! Any chance of this being merged @yardnsm?

Also, do you think it's worth updating the permissions of the temporary files so they can't be read by other system users?
